packages/flutter_tools/lib/src/plugins.dart

```diff
@@ -174,6 +174,125 @@ class Plugin {
+    if (pluginConfig == null) {
+      throwToolExit('Did not find valid `plugin` map in pubspec.yaml', exitCode: 2);
+    }
+    if (pluginConfig['platforms'] == null) {
+      throwToolExit('Did not find valid `platforms` map in pubspec.yaml', exitCode: 2);
+    }
+
+    return pluginConfig['platforms'] as YamlMap;
+  }
+
+  /// Updates the pubspec.yaml under `projectDir`, adding each platform in `platforms` under `plugin: platforms:`
+  ///
+  /// If the dummy platform `some_platform` exists, remove it.
+  /// For example, if the original `plugin` map is:
+  /// ```
+  ///   flutter:
+  ///      plugin:
+  ///        platforms:
+  ///          some_platform:
+  ///            pluginClass: somePluginClass
+  /// ```
+  /// and the `platforms` is specified as `['ios','android'], the yaml map becomes:
+  /// ```
+  ///    flutter:
+  ///      plugin:
+  ///        platforms:
+  ///          android:
+  ///            package: <androidIdentifier>
+  ///            pluginClass: <pluginClass>
+  ///          ios:
+  ///            pluginClass: <pluginClass>
+  /// ```
+  ///
+  /// This method doesn't support the legacy plugin pubspec format and will throw error if try to update a pubspec.yaml that uses the legacy format.
+  static Future<void> updatePubspecWithPlatforms(String projectDir, List<String> platforms, String pluginClass, String androidIdentifier) async {
+    final String pubspecPath = globals.fs.path.join(projectDir, 'pubspec.yaml');
+    final YamlMap pubspec = loadYaml(globals.fs.file(pubspecPath).readAsStringSync()) as YamlMap;
+    final List<String> errors = validatePluginYaml(pubspec);
+    if (errors.isNotEmpty) {
+      throwToolExit('Invalid plugin specification: \n${errors.join('\n')}');
+    }
+    try {
+      // The format of the updated pubspec might not be preserved.
+      final List<String> platformsToAdd = List<String>.from(platforms);
+      final YamlMap platformsMap = getPlatformsYamlMap(pubspec);
+      final List<String> existingPlatforms = platformsMap.keys.cast<String>().toList();
+      platformsToAdd.removeWhere((String platform) => existingPlatforms.contains(platform));
+      if (platformsToAdd.isEmpty) {
+        return;
+      }
+      final File pubspecFile = globals.fs.file(pubspecPath);
+      final List<String> fileContents = pubspecFile.readAsLinesSync();
+      int index = -1;
+      int fakePlatformIndex = -1;
+      String frontSpaces;
+      for (int i = 0; i < fileContents.length; i ++) {
+        // Find the line of `platforms:`
+        final String line = fileContents[i];
+        if (line.contains('platforms:')) {
+          final String lastLine = fileContents[i-1];
+          if (!lastLine.contains('plugin:')) {
+            continue;
+          }
+          // Find how many spaces are in front of the `platforms`.
+          frontSpaces = line.split('platforms:').first;
+          index = i + 1;
+        }
+        if (line.contains('some_platform:')) {
```

Comment:
Does this rely on on having a fake platform key?

Reply:
This logic doesn't. however it does remove the `some_platform`.
The reason that `some_platform` needs to be there is that we need to keep the pubspec.yaml valid. Otherwise, other steps of the `create` command would fail (pub get for example)

Another way is to perhaps not having the whole `flutter` map in the pubspec if there are no `--platforms` specified, then add the `flutter` map programmatically when `--platforms` is specified.

Reply:
This seems pretty fragile, what if someone develops android/ios and then wants to use add-platform to add a new desktop or web implementation?

Reply:
This should also work as the code is to look for the `platforms:` key and add an entry under it. i think I have a test covers this scenario: 'create a plugin with ios, then add macos'.
It only removes 'some_platform' map, which is an invalid platform.